lnrpc-ideal/rpc.proto

```diff
@@ -0,0 +1,809 @@
+
+    /// Block height at which this htlc was accepted.
+    int32 accept_height = 4;
+
+    /// Time at which this htlc was accepted.
+    int64 accept_time = 5;
+
+    /// Time at which this htlc was settled or canceled.
+    int64 resolve_time = 6;
+
+    /// Block height at which this htlc expires.
+    int32 expiry_height = 7;
+
+    /// Current state the htlc is in.
+    InvoiceHTLCState state = 8;
+
+    /// Custom tlv records.
+    map<uint64, bytes> custom_records = 9;
+
+    /// The total amount of the mpp payment in msat.
+    uint64 mpp_total_amt_msat = 10;
+}
+message PaymentHash {
+    /**
+    The hex-encoded payment hash of the invoice to be looked up. The passed
+    payment hash must be exactly 32 bytes, otherwise an error is returned.
+    Deprecated now that the REST gateway supports base64 encoding of bytes
+    fields.
+    */
+    string r_hash_str = 1 [ deprecated = true ];
+  
+    /**
+    The payment hash of the invoice to be looked up. When using REST, this field
+    must be encoded as base64.
+    */
+    bytes r_hash = 2;
+  }
+  
+  message ListInvoiceRequest {
+    /**
+    If set, only invoices that are not settled and not canceled will be returned
+    in the response.
+    */
+    bool pending_only = 1;
+  
+    /**
+    The index of an invoice that will be used as either the start or end of a
+    query to determine which invoices should be returned in the response.
+    */
+    uint64 index_offset = 4;
+  
+    /// The max number of invoices to return in the response to this query.
+    uint64 num_max_invoices = 5;
+  
+    /**
+    If set, the invoices returned will result from seeking backwards from the
+    specified index offset. This can be used to paginate backwards.
+    */
+    bool reversed = 6;
+  }
+  message ListInvoiceResponse {
+    /**
+    A list of invoices from the time slice of the time series specified in the
+    request.
+    */
+    repeated Invoice invoices = 1;
+  
+    /**
+    The index of the last item in the set of returned invoices. This can be used
+    to seek further, pagination style.
+    */
+    uint64 last_index_offset = 2;
+  
+    /**
+    The index of the last item in the set of returned invoices. This can be used
+    to seek backwards, pagination style.
+    */
+    uint64 first_index_offset = 3;
+  }
+  
+  message InvoiceSubscription {
+    /**
+    If specified (non-zero), then we'll first start by sending out
+    notifications for all added indexes with an add_index greater than this
+    value. This allows callers to catch up on any events they missed while they
+    weren't connected to the streaming RPC.
+    */
+    uint64 add_index = 1;
+  
+    /**
+    If specified (non-zero), then we'll first start by sending out
+    notifications for all settled indexes with an settle_index greater than
+    this value. This allows callers to catch up on any events they missed while
+    they weren't connected to the streaming RPC.
+    */
+    uint64 settle_index = 2;
```

Comment:
`add_index` and `settle_index` are used to control the point from where invoice events are returned. If those indices aren't the latest, a part of the backlog is streamed to the client first.

For the invoice states `Canceled` and `Accepted` there isn't such an index.

Reply:
Would it be better to use a last modified timestamp as the starting point? If yes, this idea could also be applied to listing invoices and payments.

Reply:
What about a hodl invoice that has been sitting in accepted for a week? I think we want all active invoices/payments (ie unresolved) rather than all _recently_ active ones (bc some can be unresolved but just sitting there).

Reply:
I think the timestamp (or global update counter which comes down to the same thing) is most useful for applications that have their own state and need to catch up after restart.

I see two advantages of using a timestamp:
* No coordination needed to determine the next counter value
* Invoices already store quite a few timestamps. On the invoice level, but also for the htlcs paying to the invoice. We could consider the max of all those timestamps, the last modified timestamp. So no new field or db migration needed. If there are modifications to the invoice that don't have an associated update of a timestamp field, those may falsely be replayed after startup. May not be a big deal, as long as the bulk of the updates isn't replayed.